### ISSUE TITLE
Microsoft.Bot.Builder.Integration.AspNet.Core 4.15 has a bug with App Service extension

### ISSUE THREAD

**Issue by woasishen** [microsoft/botbuilder-dotnet#6199]
## Version
4.15

## Describe the bug
I follow the doc [Configure .NET bot for extension](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-channel-directline-extension-net-bot?view=azure-bot-service-4.0),
and make [https://<your_app_service>.azurewebsites.net/.bot] works correctly. 
![image](https://user-images.githubusercontent.com/22291967/153551072-32884042-4584-4e6b-adfd-b80f18bc7cf2.png)

Then I follow the doc [Use Web Chat with the Direct Line App Service extension](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-channel-directline-extension-webchat-client?view=azure-bot-service-4.0), and I have generated the correct token.
But when using the token to connect by Web Socket,  it runs into a bug:
![image](https://user-images.githubusercontent.com/22291967/153550773-f9720477-c3ed-4302-a07d-4224484ed7de.png)
The web socket connects correctly, but there is no response.

When I change my <Microsoft.Bot.Builder.Integration.AspNet.Core package> to version 4.14, It works correctly~

## Expected behavior
![image](https://user-images.githubusercontent.com/22291967/153551244-39b892f1-c275-4cad-b78d-1be1e847f070.png)

Note: There is another issue, when I use 4.15, and make [https://<your_app_service>.azurewebsites.net/.bot] correctly. But some times later(a few hours), the ib & ob in  [https://<your_app_service>.azurewebsites.net/.bot] become false weirdly~
And this issue does not occur in 4.14 too.

**Comment by anishprasad01** [comment]
I am looking into this. In the meantime, the `ib` and `ob` being false could indicate that the bot is either unreachable due to a config issue, or the bot is not running. You could try setting the App Service to never idle and see if that fixes the problem.

**Comment by anishprasad01** [comment]
Did that work?

**Comment by anishprasad01** [comment]
Closing this issue due to inactivity. If you still require assistance, please comment and we can reopen the issue.

**Comment by geo-msft** [comment]
I can still repro this issue here https://clientforbotextension.azurewebsites.net/

**Comment by anishprasad01** [comment]
@geo-msft,

I'm a bit confused, the bot you linked to responds correctly.

![image](https://user-images.githubusercontent.com/33945547/156012698-f6015117-cd87-4536-931d-e726554c9027.png)

**Comment by woasishen** [comment]
I have build a test case. https://ms.portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/8d5ceb22-44c7-4367-bfc6-07378096e055/resourceGroups/cfyASEV415-group/overview

**Comment by woasishen** [comment]
![image](https://user-images.githubusercontent.com/22291967/156105704-8e4e4bdd-c8cf-48f2-ad09-1623876cfa5b.png)

.bot works well, but DL-ASE do not work

**Comment by woasishen** [comment]
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <title>Web Chat: Full-featured bundle with Direct Line Speech channel</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <script
      crossorigin="anonymous"
      src="https://cdn.botframework.com/botframework-webchat/latest/webchat-minimal.js"
    ></script>
    <style>
      html,
      body {
        background-color: #f7f7f7;
        height: 100%;
      }

      body {
        margin: 0;
      }

      #webchat {
        box-shadow: 0 0 10px rgba(0, 0, 0, 0.05);
        height: 100%;
        margin: auto;
        max-width: 480px;
        min-width: 360px;
      }
    </style>
  </head>
  <body>
    <div id="webchat" role="main"></div>
    <script>
      (async function() {
        const res = await fetch('https://cfyASEV415-webapp.azurewebsites.net/.bot/v3/directline/tokens/generate', 
{
  "method": "POST",
  "headers": {
    "Authorization": "Bearer 9fh0GVlOyoM.nDKauVVqz7hhZlrfG4wHna2R47m8cQjOTetleCLG9_I"
  },
  "body": "{'user': {  'id': 'cfyid',  'name': 'cfyname'}}"

}
);
        const { token } = await res.json();

        window.WebChat.renderWebChat(
          {
            directLine: await window.WebChat.createDirectLineAppServiceExtension({
              domain: 'https://cfyASEV415-webapp.azurewebsites.net/.bot/v3/directline',
              token
            })
          },
          document.getElementById('webchat')
        );

        document.querySelector('#webchat > *').focus();
      })().catch(err => console.error(err));
    </script>
  </body>
</html>

**Comment by geo-msft** [comment]
Hi, you can speak to my bot here https://clientforbotextension.azurewebsites.net/  because I downgraded the SDK from 4.15 to 4.14.4. If in the bot code, I update the NuGet packages to 4.15, the bot no longer replies when tested in that webchat. I took the code for the webchat from here https://docs.microsoft.com/en-us/azure/bot-service/bot-service-channel-directline-extension-webchat-client?view=azure-bot-service-4.0

**Comment by anishprasad01** [comment]
@woasishen, @geo-msft,

I attempted a reproduction myself, and I was able to get it to work. I deployed a simple Echo Bot running on 4.15.2 to Azure and set up Direct Line App Service Extension and was able to talk to the bot using the Web Chat from the linked documentation. 

I supposed the next step would be to check various points:

- AppId, Password, DirectLineExtensionKey, and other secrets. In both the bot code and in the App Service, are all of the parameters correct?
- App Service set to always run? (Check that the bot is running)
- Bot built in release mode before deployment?
- Does restarting the App Service help?
- Is the token acquisition method correct?

One more thing to try. In `Startup.cs`, add the audience in the named pipes call. Use this line instead:
```cs
app.UseNamedPipes(System.Environment.GetEnvironmentVariable("APPSETTING_WEBSITE_SITE_NAME") + ".directline", "https://api.botframework.com");
```

Then redeploy the bot and try again.

**Comment by anishprasad01** [comment]
@woasishen,

I deleted one of your comments because your Direct Line secret was visible. This is a public repository, so please be careful with what you post.

**Comment by woasishen** [comment]
@anishprasad01 
Thanks. You are right, it is not suitable to put the secret here.
I have changed the app.UseNamedPipes, and redeploy & restart.  But the question still there, 

The token acquisition method correct, and get the right token. 
![image](https://user-images.githubusercontent.com/22291967/156690784-2cb0482e-9109-45fc-9fa8-09736891a672.png)
![image](https://user-images.githubusercontent.com/22291967/156690842-e64109fa-26df-4796-bb87-49695592aa6c.png)
![image](https://user-images.githubusercontent.com/22291967/156690968-b91d29b3-ab8d-4930-b341-bc0b31a49189.png)
![image](https://user-images.githubusercontent.com/22291967/156690991-cd6df447-99e6-4003-93b4-0a203bde83e8.png)
![image](https://user-images.githubusercontent.com/22291967/156691035-3e1a8531-142a-4a16-b2ad-b2a239d1079b.png)
![image](https://user-images.githubusercontent.com/22291967/156691066-46054a23-9352-43bb-b023-083648807edf.png)

**Comment by woasishen** [comment]
@anishprasad01 
It seems everything works well, except the last picture.
And if I change the <microsoft.bot.builder.integration.aspnet.core> version from 4.15.2 to 4.14.2. The issue in the last picture will be not exist.

**Comment by woasishen** [comment]
![image](https://user-images.githubusercontent.com/22291967/156691628-2c39b14b-4001-4e99-9626-ce1e444867fd.png)

**Comment by anishprasad01** [comment]
Have you configured CORS settings in your App Service? for testing purposes, you can just use `*` to allow all origins.

**Comment by woasishen** [comment]
Last time I don't know CORS. Now I try add CORS, But still not work the same.
![image](https://user-images.githubusercontent.com/22291967/156881363-91ed14f3-77cd-422d-881e-f6f5a84ca70a.png)

**Comment by anishprasad01** [comment]
I took another look at your web app, and I noticed that it does not have the `MicrosoftAppType`, `MicrosoftAppId`, `MicrosoftAppPassword`, and other fields that would be expected to be present when running a published bot. How exactly did you publish your bot?

**Comment by woasishen** [comment]
![image](https://user-images.githubusercontent.com/22291967/157493308-f4f0a84c-5bc0-4b29-bdbb-85eeb3d20c56.png)

you mean this?

**Comment by anishprasad01** [comment]
No, I'm referring to Application Settings in the App Service Configuration blade.

![image](https://user-images.githubusercontent.com/33945547/157495780-f0a9de0e-caaa-4b4e-8175-3642fce871a9.png)

When you deploy a bot using the Azure CLI using the steps from the [documentation](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-deploy-az-cli?view=azure-bot-service-4.0), you should end up with fields here defining these parameters for your App Service. That's why I asked how you deployed your bot.

**Comment by woasishen** [comment]
![image](https://user-images.githubusercontent.com/22291967/157789349-780651f6-c8c1-42ed-929c-15403794f0e4.png)
![image](https://user-images.githubusercontent.com/22291967/157789412-7ca20796-7b26-44ec-b450-da72d45c0b5d.png)

I just create a new proj use <echo bot .net core3.1>
![image](https://user-images.githubusercontent.com/22291967/157789572-2a0c941a-9490-4dd9-97a6-91590bf93fa8.png)
And add UseNamedPipes, Change appsetting. Then publish by visualstudio.
![image](https://user-images.githubusercontent.com/22291967/157789763-3555b554-6b1d-4cff-b461-cecf379a944c.png)
![image](https://user-images.githubusercontent.com/22291967/157789864-02635a1b-ef44-43da-8605-b5bc3f148a9b.png)

So it is realy easy to reproduce.
And I have update my test client html.
If you follow my step, you may reproduce my bug.
I will upload my html below, this is just copy from your document.<https://docs.microsoft.com/en-us/azure/bot-service/bot-service-channel-directline-extension-webchat-client?view=azure-bot-service-4.0>

**Comment by woasishen** [comment]
<!DOCTYPE html>
<html lang="en-US">
  <head>
    <title>Web Chat: Full-featured bundle with Direct Line Speech channel</title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <script
      crossorigin="anonymous"
      src="https://cdn.botframework.com/botframework-webchat/latest/webchat-minimal.js"
    ></script>
    <style>
      html,
      body {
        background-color: #f7f7f7;
        height: 100%;
      }

      body {
        margin: 0;
      }

      #webchat {
        box-shadow: 0 0 10px rgba(0, 0, 0, 0.05);
        height: 100%;
        margin: auto;
        max-width: 480px;
        min-width: 360px;
      }
    </style>
  </head>
  <body>
    <div id="webchat" role="main"></div>
    <script>
      (async function() {
        const res = await fetch('https://cfyASEV415-webapp.azurewebsites.net/.bot/v3/directline/tokens/generate', 
{
  "method": "POST",
  "headers": {
    "Authorization": "Bearer 9fh0GVl...<DirectLine Secret>"
  },
  "body": "{'user': {  'id': 'cfyid',  'name': 'cfyname'}}"

}
);
        const { token } = await res.json();

        window.WebChat.renderWebChat(
          {
            directLine: await window.WebChat.createDirectLineAppServiceExtension({
              domain: 'https://cfyASEV415-webapp.azurewebsites.net/.bot/v3/directline',
              token
            })
          },
          document.getElementById('webchat')
        );

        document.querySelector('#webchat > *').focus();
      })().catch(err => console.error(err));
    </script>
  </body>
</html>

**Comment by anishprasad01** [comment]
After taking another look at you App Service in Azure, It looks like there are residual files and for some reason your `Bot.Builder.Integration` (the key package in question) version is set to `4.14.0-local`.

At this point, I think you should try a clean wipe and deploy the bot fresh. I will lay out the steps I used to get my bot working, and hopefully by following them you can get it working as well.

1. Create a simple echo bot by copying the sample code.
2. Add the named pipes and websockets commands as per the [documentation](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-channel-directline-extension-net-bot?view=azure-bot-service-4.0).
3. Add the "https://api.botframework.com" audience to the named pipes call (you already did this I believe).
4. Update Bot Builder packages to 4.15.2
5. Clean and Rebuild in Release mode
6. Follow [docs](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-deploy-az-cli?view=azure-bot-service-4.0) to deploy the bot to Azure via the Azure CLI.
7. Follow the DL-ASE [document](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-channel-directline-extension-net-bot?view=azure-bot-service-4.0#enable-bot-direct-line-app-service-extension) to enable the App Service Extension and verify `.bot` works.
8. Follow the documentation to [get a token](https://docs.microsoft.com/en-us/azure/bot-service/bot-service-channel-directline-extension-webchat-client?view=azure-bot-service-4.0) using something like Postman. Make sure you are sending your secret as a Bearer token.
9. Get the example web chat for DL-ASE from before. Since we are just testing locally, remove the lines for fetching the token from a token server.
10. Place your token below the domain in the designated place. Follow the same formatting.
11. Save the file and open it.
12. You should now be able to speak to the bot.

**Comment by anishprasad01** [comment]
@woasishen, @geo-msft,

I created a new bot and web chat and published them to Azure, and can now reproduce and confirm this issue. With the repro in hand, we now have a path forward to investigate this issue. Apologies for the delay, and thanks for your patience.

**Comment by anishprasad01** [comment]
@geo-msft, @woasishen

It appears that there is some sort of bug in the `CloudAdapter` in SDK 4.15 that prevents ASE from working correctly with the new streaming library introduced in 4.15. The named pipes functionality required for ASE was [recently updated](https://github.com/microsoft/botbuilder-dotnet/pull/6147) to use the new streaming library, thereby fixing this issue. However, the fix is not currently available in any stable build. 

**The fix will be part of the upcoming SDK 4.16 release**. 

Thus, upgrading to 4.16 should resolve this issue. It is possible to test the fix now by getting the [daily build of the SDK from MyGet](https://github.com/microsoft/botbuilder-dotnet/blob/main/UsingMyGet.md), and I can confirm that the daily build of 4.16.0 does appear to work with ASE. For production scenarios, customers should wait for the stable release of 4.16 coming in the near future.

**Comment by anishprasad01** [comment]
Since we have identified a workaround for testing and development in the short-term (the daily builds) and a long-term fix (SDK 4.16), I am going to go ahead and close this issue.
